src/compiler/checker.ts

```diff
@@ -32023,14 +32045,46 @@ namespace ts {
         }
 
         function checkInExpression(left: Expression, right: Expression, leftType: Type, rightType: Type): Type {
-            if (leftType === silentNeverType || rightType === silentNeverType) {
-                return silentNeverType;
+            if (isPrivateIdentifier(left)) {
+                const lexicallyScopedSymbol = lookupSymbolForPrivateIdentifierDeclaration(left.escapedText, left);
+                if (lexicallyScopedSymbol === undefined) {
+                    if (!getContainingClass(left)) {
+                        error(left, Diagnostics.Private_identifiers_are_not_allowed_outside_class_bodies);
+                    }
+                    else {
+                        const suggestion = getSuggestedSymbolForNonexistentProperty(left, rightType);
+                        if (suggestion) {
+                            const suggestedName = symbolName(suggestion);
+                            error(left, Diagnostics.Cannot_find_name_0_Did_you_mean_1, diagnosticName(left), suggestedName);
+                        }
+                        else {
+                            error(left, Diagnostics.Cannot_find_name_0, diagnosticName(left));
+                        }
+                    }
+                    return anyType;
+                }
+
+                markPropertyAsReferenced(lexicallyScopedSymbol, /* nodeForCheckWriteOnly: */ undefined, /* isThisAccess: */ false);

```

Comment:
I've added a `checkPrivateIdentifierExpression`, and added it to the `checkExpression` path. So the PrivateId is checked on it's own as usual for the LHS of a binary-expression, and also when we have parsed it in invalid situations. This function is what sets the `resolvedSymbol`, akin to a regular identifier.
I named it differently because it's not being used in `checkPrivateIdentifierPropertyAccess` - I wanted to make it clearer that it is only called when we are checking the PrivateId as an expression.
Though I was able to increase some of the shared code by switching to `reportNonexistentProperty`  which _is_ shared by `checkPropertyAccessExpressionOrQualifiedName`.

There is still some added logic to `checkInExpression`, as at this point we now have the RHS type too.